src/main/java/org/terasology/gookeeper/system/EconomySystem.java

```diff
@@ -75,32 +93,120 @@
+            String message = blockCommands.giveBlock(client, itemPrefabName, amount, shapeUriParam);
+            if (message != null) {
+                economyComponent.playerWalletCredit -= quantityParam * Assets.getPrefab(itemPrefabName + "Fenced").get().getComponent(PurchasableComponent.class).basePrice;
+                player.saveComponent(economyComponent);
+                return "Successfully purchased " + itemPrefabName;
+            } else {
+                return "Couldn't find requested block.";
+            }
+        } else {
+            return "You dont have sufficient balance to purchase.";
+        }
+    }
+
+    /**
+     * This command can be used to purchase utility blocks (such as pen blocks, visit blocks, etc.)
+     *
+     * @param client
+     * @param itemPrefabName
+     * @param amount
+     * @param shapeUriParam
+     * @return
+     */
+    @Command(shortDescription = "Purchase utility blocks",
+            requiredPermission = PermissionManager.NO_PERMISSION)
+    public String upgrade(@Sender EntityRef client,
+                           @CommandParam("prefabId or blockName") String itemPrefabName) {
+        EntityRef player = client.getComponent(ClientComponent.class).character;
+        EconomyComponent economyComponent = player.getComponent(EconomyComponent.class);
+
+        for (int i = 0; i < inventoryManager.getNumSlots(player); i++) {
+            EntityRef itemInSlot = inventoryManager.getItemInSlot(player, i);
+            DisplayNameComponent displayNameComponent = itemInSlot.getComponent(DisplayNameComponent.class);
+
+            if (displayNameComponent != null && displayNameComponent.name.equals(itemPrefabName)) {
+                UpgradableComponent upgradableComponent = itemInSlot.getComponent(UpgradableComponent.class);
+                if (upgradableComponent != null && (economyComponent.playerWalletCredit - (upgradableComponent.baseUpgradePrice * upgradableComponent.currentTier) > 0f)) {
+                    upgradeByType(itemInSlot, upgradableComponent);
+                    economyComponent.playerWalletCredit -= upgradableComponent.baseUpgradePrice * upgradableComponent.currentTier;
+                    player.saveComponent(economyComponent);
+                    itemInSlot.saveComponent(upgradableComponent);
+                    return "Successfully upgraded your " + displayNameComponent.name;
+                } else {
+                    return "You dont have sufficient balance to upgrade the item.";
                 }
             }
         }
+
+        return "Couldn't find the requested item to upgrade.";
+    }
+
+    private void upgradeByType (EntityRef item, UpgradableComponent upgradableComponent) {
+        upgradableComponent.currentTier ++;
+
+        if (item.hasComponent(SlimePodItemComponent.class)) {
```

Comment:
This method has a few code smells, because it's very rigid and won't let you extend the upgrading to new blocks/items very easily.
However, I say we should open a TODO issue to improve it and then unify the upgrading code between GooeyDefence and this later

Reply:
Yeah currently it is hard-coded to just work with the two items, and hence definitely not the ideal way to do it. 
Yeah, a unified upgrade module seems great 😄